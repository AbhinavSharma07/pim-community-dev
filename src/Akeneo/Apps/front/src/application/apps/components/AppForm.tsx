import * as React from 'react';

export const AppForm = () => <>AppForm</>;
