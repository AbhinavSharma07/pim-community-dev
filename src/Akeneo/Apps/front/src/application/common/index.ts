import {Breadcrumb} from './breadcrumb/Breadcrumb';
import {Item as BreadcrumbItem} from './breadcrumb/Item';
import {ApplyButton} from './button/ApplyButton';
import {Header} from './Header';
import {Page} from './Page';

export {Header, Page, Breadcrumb, BreadcrumbItem, ApplyButton};
