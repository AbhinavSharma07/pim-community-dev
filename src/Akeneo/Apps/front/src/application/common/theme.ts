import {ThemedStyledProps} from 'styled-components';

export const theme = {
    color: {
        blue: '#5992c7',
        darkBlue: '#11324d',
        grey: '#a1a9b7',
        slateGrey: '#67768a',
    },
};

export type PropsWithTheme<P = {}> = ThemedStyledProps<P, typeof theme>;
