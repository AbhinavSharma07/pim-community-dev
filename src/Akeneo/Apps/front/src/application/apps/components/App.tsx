import React from 'react';
import {useHistory} from 'react-router';
import {Figure, FigureImage, FigureCaption} from '../../common';
import imgUrl from '../../common/assets/illustrations/Api.svg';

interface Props {
    code: string;
    label: string;
}

export const App = ({code, label}: Props) => {
    const history = useHistory();

    return (
        <Figure onClick={() => history.push(`/apps/${code}/edit`)}>
            <FigureImage src={imgUrl} alt={label} />
            <FigureCaption title={label}>{label}</FigureCaption>
        </Figure>
    );
};
