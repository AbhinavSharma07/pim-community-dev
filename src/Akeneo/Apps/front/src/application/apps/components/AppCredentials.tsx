import React, {FC, useContext} from 'react';
import {AppCredentials as AppCredentialsInterface} from '../../../domain/apps/app-credentials.interface';
import {InlineHelper, Section, SmallHelper, HelperLink} from '../../common';
import {Translate, TranslateContext} from '../../shared/translate';
import {CopiableCredential} from './credentials/CopiableCredential';
import {Credential, CredentialList} from './credentials/Credential';
import {RegenerateSecretButton} from './RegenerateSecretButton';

type Props = {
    code: string;
    label: string;
    credentials: AppCredentialsInterface;
};

export const AppCredentials: FC<Props> = ({code, label, credentials: credentials}: Props) => {
    const translate = useContext(TranslateContext);

    return (
        <>
            <Section title={<Translate id='akeneo_apps.edit_app.credentials.title' />} />
            <div>
                <SmallHelper>
                    <Translate id='akeneo_apps.edit_app.credentials.helper.message' placeholders={{label}} />
                    &nbsp;
                    <HelperLink
                        href={translate('akeneo_apps.edit_app.credentials.helper.link_url')}
                        target='_blank'
                        rel='noopener noreferrer'
                    >
                        <Translate id='akeneo_apps.edit_app.credentials.helper.link' />
                    </HelperLink>
                </SmallHelper>
            </div>

            <CredentialList>
                <CopiableCredential label={translate('akeneo_apps.app.client_id')}>
                    {credentials.clientId}
                </CopiableCredential>
                <CopiableCredential
                    label={translate('akeneo_apps.app.secret')}
                    actions={<RegenerateSecretButton code={code} />}
                >
                    {credentials.secret}
                </CopiableCredential>
                <CopiableCredential label={translate('akeneo_apps.app.username')}>
                    {credentials.username}
                </CopiableCredential>
                {credentials.password ? (
                    <CopiableCredential
                        label={translate('akeneo_apps.app.password')}
                        helper={
                            <InlineHelper warning>
                                <Translate id='akeneo_apps.edit_app.credentials.clear_password_helper.message' />{' '}
                                <a
                                    href={translate('akeneo_apps.edit_app.credentials.clear_password_helper.link_url')}
                                    target='_blank'
                                    rel='noopener noreferrer'
                                >
                                    <Translate id='akeneo_apps.edit_app.credentials.clear_password_helper.link' />
                                </a>
                            </InlineHelper>
                        }
                    >
                        {credentials.password}
                    </CopiableCredential>
                ) : (
                    <Credential
                        label={translate('akeneo_apps.app.password')}
                        helper={
                            <InlineHelper info>
                                <Translate id='akeneo_apps.edit_app.credentials.password_helper.message' />{' '}
                                <a
                                    href={translate('akeneo_apps.edit_app.credentials.password_helper.link_url')}
                                    target='_blank'
                                    rel='noopener noreferrer'
                                >
                                    <Translate id='akeneo_apps.edit_app.credentials.password_helper.link' />
                                </a>
                            </InlineHelper>
                        }
                    >
                        <Translate id='akeneo_apps.edit_app.credentials.password_placeholder' />
                    </Credential>
                )}
            </CredentialList>
        </>
    );
};
