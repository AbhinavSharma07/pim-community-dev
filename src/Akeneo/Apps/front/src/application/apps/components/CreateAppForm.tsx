import React, {ChangeEvent, useRef, useEffect, useContext, useReducer} from 'react';
import {FlowType} from '../../../domain/apps/flow-type.enum';
import {ApplyButton, Form, FormGroup, FormInput} from '../../common';
import {FlowTypeHelper} from './FlowTypeHelper';
import {FlowTypeSelect} from './FlowTypeSelect';
import {appFormReducer, CreateFormState} from '../reducers/app-form-reducer';
import {inputChanged, setError, setValidated} from '../actions/create-form-actions';
import {sanitize} from '../../shared/sanitize';
import {Translate, TranslateContext} from '../../shared/translate';
import {fetch} from '../../shared/fetch';
import {isErr} from '../../shared/fetch/result';
import {NotificationLevel, useNotify} from '../../shared/notify';
import {useRoute} from '../../shared/router';
import {useHistory} from 'react-router';

const initialState: CreateFormState = {
    controls: {
        code: {name: 'code', value: '', errors: {}, dirty: false, valid: false, validated: true},
        label: {name: 'label', value: '', errors: {}, dirty: false, valid: false, validated: true},
        flow_type: {
            name: 'flow_type',
            value: FlowType.DATA_SOURCE,
            errors: {},
            dirty: false,
            valid: true,
            validated: true,
        },
    },
    valid: false,
};

interface ResultError {
    message: string;
    errors: Array<{
        name: string;
        reason: string;
    }>;
}

export const CreateAppForm = () => {
    const [state, dispatch] = useReducer(appFormReducer, initialState);

    const codeInputRef = useRef<HTMLInputElement>(null);
    const labelInputRef = useRef<HTMLInputElement>(null);

    const url = useRoute('akeneo_apps_create_rest');
    const notify = useNotify();
    const translate = useContext(TranslateContext);
    const history = useHistory();

    useEffect(() => {
        [codeInputRef, labelInputRef].forEach(inputRef => {
            const input = inputRef.current;
            if (null === input) {
                return;
            }

            const name = input.name;
            // if (true === state.controls[name].validated) {
            //     return;
            // }

            if (false === input.checkValidity()) {
                if (input.validity.valueMissing) {
                    dispatch(setError(name, `akeneo_apps.constraint.${name}.required`));
                }
                if (input.validity.patternMismatch) {
                    dispatch(setError(name, `akeneo_apps.constraint.${name}.invalid`));
                }
            }
            // dispatch(setValidated(name));
        });
    }, [state.controls.code.value, state.controls.label.value, dispatch]);

    useEffect(() => {
        if (true === state.controls.code.dirty) {
            return;
        }

        const value = sanitize(state.controls.label.value);
        if (state.controls.code.value === value) {
            return;
        }

        dispatch(inputChanged('code', value, false));
    }, [state.controls.label.value, dispatch]);

    const handleSave = async () => {
        if (false === state.valid) {
            return;
        }
        const data = {
            code: state.controls.code.value,
            label: state.controls.label.value,
            flow_type: state.controls.flow_type.value,
        };

        const result = await fetch<undefined, ResultError>(url, {
            method: 'POST',
            headers: [['Content-type', 'application/json']],
            body: JSON.stringify(data),
        });

        if (isErr(result)) {
            if (undefined === result.error.errors) {
                notify(NotificationLevel.ERROR, translate('pim_apps.create_app.flash.error'));
                return;
            }
            result.error.errors.forEach(({name, reason}) => dispatch(setError(name, reason)));
            return;
        }

        notify(NotificationLevel.SUCCESS, translate('pim_apps.create_app.flash.success'));
        history.push(`/apps/${data.code}/edit`);
    };

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
        dispatch(inputChanged(event.currentTarget.name, event.currentTarget.value));
    };

    const handleFlowTypeSelect = (flowType: FlowType) => dispatch(inputChanged('flow_type', flowType));

    return (
        <Form>
            <FormGroup controlId='label' label='pim_apps.app.label' errors={Object.keys(state.controls.label.errors)}>
                <FormInput
                    ref={labelInputRef}
                    type='text'
                    name='label'
                    value={state.controls.label.value}
                    onChange={handleChange}
                    required
                    maxLength={100}
                />
            </FormGroup>

            <FormGroup controlId='code' label='pim_apps.app.code' errors={Object.keys(state.controls.code.errors)}>
                <FormInput
                    ref={codeInputRef}
                    type='text'
                    name='code'
                    value={state.controls.code.value}
                    onChange={handleChange}
                    required
                    maxLength={100}
                    pattern='^[0-9a-zA-Z_]+$'
                />
            </FormGroup>

            <FormGroup
                controlId='flow_type'
                label='pim_apps.app.flow_type'
                info={<FlowTypeHelper />}
                required
                errors={Object.keys(state.controls.flow_type.errors)}
            >
                <FlowTypeSelect value={state.controls.flow_type.value as FlowType} onChange={handleFlowTypeSelect} />
            </FormGroup>

            <ApplyButton onClick={handleSave} disabled={false === state.valid}>
                <Translate id='pim_common.save' />
            </ApplyButton>
        </Form>
    );
};
