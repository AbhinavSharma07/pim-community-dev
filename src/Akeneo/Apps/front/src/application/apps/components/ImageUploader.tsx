import React, {useContext} from 'react';
import {useImageUploader} from '../use-image-uploader';
import {useMediaUrlGenerator} from '../use-media-url-generator';
import {Translate, TranslateContext} from '../../shared/translate';
import styled from 'styled-components';
import {PropsWithTheme} from '../../common/theme';
import Trash from '../../common/assets/icons/trash';
import {isErr, isOk} from '../../shared/fetch/result';
import {NotificationLevel, useNotify} from '../../shared/notify';
import defaultImageUrl from '../../common/assets/illustrations/api.svg';

interface Props {
    image: string|null;
    onChange: (image: string|null) => void;
    onError: (error: string) => void;
}

const HelperLink = styled.a`
    color: ${({theme}: PropsWithTheme) => theme.color.blue};
    text-decoration: underline;
    font-weight: 700;
`;

const DefaultImage = styled.img`
    max-height: 140px;
    width: auto;
`;

const ImageUploader = ({image, onChange, onError}: Props) => {
    const containerClassName = `AknImage AknImage--editable AknImage--wide ${null === image ? 'AknImage--empty' : ''}`;
    const imageUploader = useImageUploader();
    const generateMediaUrl = useMediaUrlGenerator();
    const notify = useNotify();
    const translate = useContext(TranslateContext);

    const upload = async (file: File) => {
        const result = await imageUploader(file);
        if (isErr(result)) {
            if (undefined !== result.error.extension) {
                notify(NotificationLevel.ERROR, translate('pim_apps.edit_image.flash.extension_not_allowed'));
            } else {
                const errors = Object.entries(result.error).reduce((errors, [propertyPath, {message}]) => {
                    return errors + `${propertyPath}: ${message}`;
                }, '');
                onError(errors);
            }

            return null;
        }

        if (isOk(result)) {
            return result.data.filePath;
        }
        return null;
    };

    const onInputChange = async (event: React.ChangeEvent<HTMLInputElement>) => {
        if (null !== event.target.files) {
            const mediaUrl = await upload(event.target.files[0]);
            if (null !== mediaUrl) {
                onChange(mediaUrl);
            }
        }
    };

    const previewImage = image ? generateMediaUrl(image) : null;

    return (
        <>
            <div className={containerClassName}>
                <input
                  type='file'
                  accept='.jpg, .jpeg, .gif, .png, .wbmp, .xbm, .webp, .bmp'
                  onChange={onInputChange}
                  className='AknImage-updater'
                />

                {null === previewImage && (
                    <div className='AknImage-uploader'>
                        <DefaultImage
                            src={defaultImageUrl}
                            alt={''}
                        />
                        <span className='AknImage-uploaderIllustration' />
                        <span className='AknImage-uploaderHelper'>
                            <Translate id={'pim_apps.edit_image.upload_helper'} />{' '}
                            <HelperLink href='#'>
                                <Translate id={'pim_apps.edit_image.click_here'} />
                            </HelperLink>.
                        </span>
                    </div>
                )}
                {null !== previewImage && (
                    <>
                        <div className='AknImage-action'>
                            <span className='AknImage-actionItem' onClick={() => onChange(null)}>
                                <Trash color='#ffffff' className='AknImage-actionItemIcon' />{' '}
                                <Translate id={'pim_apps.edit_image.remove_helper'} />
                            </span>
                        </div>
                        <div className='AknImage-displayContainer'>
                            <img
                                className='AknImage-display'
                                src={previewImage}
                                alt={''}
                            />
                        </div>
                    </>
                )}
            </div>
        </>
    );
};

export default ImageUploader;
