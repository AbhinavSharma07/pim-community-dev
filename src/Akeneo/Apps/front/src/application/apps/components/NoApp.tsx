import React from 'react';
import styled from 'styled-components';
import {PageError} from '../../common';
import {PropsWithTheme} from '../../common/theme';
import {Translate} from '../../shared/translate';

export const NoApp = ({onCreate}: {onCreate: () => void}) => (
    <PageError title={<Translate id='pim_apps.no_app.title' />}>
        <Translate id='pim_apps.no_app.message' />
        &nbsp;
        <Link onClick={onCreate}>
            <Translate id='pim_apps.no_app.message_link' />
        </Link>
    </PageError>
);

const Link = styled.a`
    color: #9452ba;
    cursor: pointer;
    text-decoration: underline ${({theme}: PropsWithTheme) => theme.color.purple100};
`;
