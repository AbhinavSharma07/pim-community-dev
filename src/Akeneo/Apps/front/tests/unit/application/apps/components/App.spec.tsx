import {mount} from 'enzyme';
import {createMemoryHistory} from 'history';
import * as React from 'react';
import {MemoryRouter, Router} from 'react-router';
import {ThemeProvider} from 'styled-components';
import {App} from '../../../../../src/application/apps/components/App';
import {theme} from '../../../../../src/application/common/theme';
import {createWithTheme} from '../../../../utils/create-with-theme';

describe('App', () => {
    it('should render', () => {
        const component = createWithTheme(
            <MemoryRouter>
                <App code={'google-shopping'} label={'Google Shopping'} />
            </MemoryRouter>
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should redirect to the app edit page when clicked', () => {
        const history = createMemoryHistory();
        const component = mount(
            <ThemeProvider theme={theme}>
                <Router history={history}>
                    <App code={'google-shopping'} label={'Google Shopping'} />
                </Router>
            </ThemeProvider>
        );

        component.simulate('click');

        expect(history.location.pathname).toBe('/apps/google-shopping/edit');
    });
});
