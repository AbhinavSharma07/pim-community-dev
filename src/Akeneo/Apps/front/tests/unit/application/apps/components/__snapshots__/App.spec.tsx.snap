// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `
<div
  className="sc-bdVaJa Dsuco"
  onClick={[Function]}
>
  <div
    className="sc-htpNat iQoyqR"
  >
    <img
      alt="Google Shopping"
      className="sc-bwzfXH foaHqu"
      src="test-file-stub"
    />
  </div>
  <div
    className="sc-bxivhb ffTZNS"
    title="Google Shopping"
  >
    Google Shopping
  </div>
</div>
`;
